src/Plutarch/Extra/Map.hs

```diff
@@ -1,185 +1,363 @@
-{-# LANGUAGE PartialTypeSignatures #-}
 {-# LANGUAGE QuantifiedConstraints #-}
-{-# LANGUAGE ViewPatterns #-}
 
+{- |
+ Module: Plutarch.Extra.Map
+ Copyright: (C) Liqwid Labs 2022
+ License: Apache 2.0
+ Maintainer: Koz Ross <koz@mlabs.city>
+ Portability: GHC only
+ Stability: Experimental
+
+ Various helpers for 'PMap' use. This module is intended to be imported
+ qualified.
+-}
 module Plutarch.Extra.Map (
-    plookup,
-    plookup',
-    pmapFromList,
-    pkeys,
+    -- * Lookup
+    ptryLookup,
+
+    -- * Comparisons
+    pkeysEqual,
+    pkeysEqualUnsorted,
+
+    -- * Modification
     pupdate,
-    Plutarch.Extra.Map.pmap,
+    padjust,
+
+    -- * Folds
+    pfoldMapWithKey,
+    pfoldlWithKey,
+
+    -- * Conversion
+    punsortedMapFromFoldable,
+    psortedMapFromFoldable,
+    pkeys,
+
+    -- * Key-value pair manipulation
     pkvPairKey,
+    pkvPairValue,
     pkvPairLt,
-    pfoldlWithKey,
-    pfoldMapWithKey,
 ) where
 
 import Data.Foldable (foldl')
-import Plutarch.Api.V1.AssocMap (KeyGuarantees, PMap (PMap))
+import Plutarch.Api.V1.AssocMap (
+    KeyGuarantees (Sorted, Unsorted),
+    PMap (PMap),
+    pdelete,
+    pempty,
+    pinsert,
+    plookup,
+ )
 import Plutarch.Builtin (ppairDataBuiltin)
-import Plutarch.Extra.Functor (pfmap)
-import Plutarch.Extra.List (pmapMaybe)
-import Plutarch.Extra.TermCont (pletC, pmatchC)
-import Plutarch.Extra.Traversable (pfoldMap)
-import qualified Plutarch.List
-
--- | @since 1.0.0
-plookup ::
-    forall (k :: S -> Type) (v :: S -> Type) (keys :: KeyGuarantees) (s :: S).
-    (PIsData v, PIsData k, PEq k) =>
-    Term s (k :--> PMap keys k v :--> PMaybe v)
-plookup = phoistAcyclic $
-    plam $ \x m -> unTermCont $ do
-        PMap kvs <- pmatchC m
-        res <- pmatchC (pfind # (go # x) # kvs)
-        pure $ case res of
-            PNothing -> pcon PNothing
-            PJust y -> pcon . PJust . pfromData $ psndBuiltin # y
+import Plutarch.Extra.Maybe (passertPJust)
+import qualified Plutarch.List as PList
+
+{- | If a value exists at the specified key, apply the function argument to it;
+ otherwise, do nothing.
+
+ This is necessarily linear in the size of the map performance-wise, as we
+ have to scan the entire map to find the key in the worst case.
+
+ @since 3.3.0
+-}
+padjust ::
+    forall (k :: S -> Type) (v :: S -> Type) (s :: S).
+    (PIsData k, PEq k, PIsData v) =>
+    Term s ((v :--> v) :--> k :--> PMap 'Unsorted k v :--> PMap 'Unsorted k v)
+padjust = phoistAcyclic $
+    plam $ \f key kvs ->
+        pmatch kvs $ \(PMap kvs') ->
+            pcon . PMap $ PList.pmap # (go # f # key) # kvs'
   where
     go ::
```

Comment:
Having Haskell function here would be cleaner/No need for `phoistAcyclic`. 

```hs
go :: Term s k -> Term s (PBuiltinPair (PAsData k) (PAsData k)) -> PBool
```

Reply:
Yeah, I agree there, especially since it'll be inlined anyway. Will fix.